TelepresenceBot/server-unit/core/server.js

```diff
@@ -1,21 +1,64 @@
 var io = require('socket.io').listen(5000);
+var ClientType = require("./clientType.js");
+var LoggingClient = require("./loggingClient.js");
 
-var clients = {};
+var room = "London";
+var humans = [];
+var bots = [];
+var testClient = new LoggingClient();
+
+io.use(function(client, next){
+    var rawClientType = client.handshake.query.clientType;
+    var clientType = ClientType.from(rawClientType);
+
+    switch(clientType) {
+        case ClientType.TEST:
+        case ClientType.BOT:
+            return next();
+        case ClientType.HUMAN:
+            var roomRoster = io.sockets.adapter.rooms[room];
+            if(roomRoster != undefined && roomRoster.length == 1) {
+                return next();
+            } else {
+                return next(new Error('No bots available'));
+            }
+        default:
+            return next(new Error('Unrecognised clientType: ' + rawClientType));
+
+    }
+});
 
 io.sockets.on('connection', function (client) {
 
-    console.log('connecting: ' + client.id);
-    clients[client.id] = client;
-    io.sockets.emit('connected', toKeysArrayFrom(clients));
+    var rawClientType = client.handshake.query.clientType;
+    var clientType = ClientType.from(rawClientType);
+
+    switch(clientType) {
+        case ClientType.HUMAN:
+            humans.push(client.id);
+            client.join(room);
+            testClient.emit('connected_human', humans);
+            break;
+        case ClientType.BOT:
+            bots.push(client.id);
+            client.join(room);
+            testClient.emit('connected_bot', bots);
+            break;
+        case ClientType.TEST:
+            console.log('switching test client');
+            testClient = new LoggingClient(client);
+            testClient.emit('connected');
+            break;
+        default:
+            throw 'Unexpected rawClientType: ' + clientType;
+    }
+
 
     client.on('disconnect', function() {
-        console.log('disconnecting: ' + client.id);
-        delete clients[client.id];
-        io.sockets.emit('disconnected', toKeysArrayFrom(clients));
+        humans.splice(humans.indexOf(client.id), 1);
+        bots.splice(bots.indexOf(client.id), 1);
+        testClient.emit('disconnected_human', humans);
+        testClient.emit('disconnected_bot', bots);
```

Comment:
I could do the check here. For the splice they should be in the correct list, if they are not then the tests will pick up on it which is why the `testClient` requests both of the lists whenever a client is disconnected rather than just the one it is interested in.